### ISSUE TITLE
bpf-framework support fails to build with gcc-bpf

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#31869]
I'm trying to build systemd with bpf-framework support.
In order to avoid the rather heavyweight clang/llvm build dependency and seeing that there is a `bpf-compiler` option, I tried installing gcc-bpf-13 from https://tracker.debian.org/pkg/gcc-bpf

This yields:
```
$ meson -Dbpf-framework=enabled -Dbpf-compiler=gcc build
...
  User defined options
    bpf-compiler                      : gcc
    bpf-framework                     : enabled

$ ninja -C build/
ninja: Entering directory `build/'
[55/2459] Generating src/core/bpf/sock...bpf.unstripped.o with a custom command
FAILED: src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mkernel=5.2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/socket_bind/socket-bind.bpf.c -o src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o
bpf-gcc: error: unrecognized command-line option ‘-mkernel=5.2’
[56/2459] Generating src/core/bpf/rest...bpf.unstripped.o with a custom command
FAILED: src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mkernel=5.2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/restrict_fs/restrict-fs.bpf.c -o src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o
bpf-gcc: error: unrecognized command-line option ‘-mkernel=5.2’
[72/2459] Generating src/shared/ethtoo...d (wrapped by meson to capture output)
ninja: build stopped: subcommand failed.
```

Dropping the build flag `-mkernel=5.2` results in

``` 
$ git diff
diff --git a/meson.build b/meson.build
index 0e5f403484..4f5fec29f9 100644
--- a/meson.build
+++ b/meson.build
@@ -1692,7 +1692,6 @@ if conf.get('BPF_FRAMEWORK') == 1
                 '-std=gnu11',
                 '-fno-stack-protector',
                 '-O2',
-                '-mkernel=5.2',
                 '-mcpu=v3',
                 '-mco-re',
                 '-gbtf',

$ ninja -C build/
ninja: Entering directory `build/'
[2/2361] Generating src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o with a custom command
FAILED: src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/socket_bind/socket-bind.bpf.c -o src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o
In file included from ../src/core/bpf/socket_bind/socket-bind-api.bpf.h:9,
                 from ../src/core/bpf/socket_bind/socket-bind.bpf.c:9:
/usr/include/linux/types.h:5:10: fatal error: asm/types.h: No such file or directory
    5 | #include <asm/types.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
[3/2361] Generating src/core/bpf/restrict_ifaces/restrict-ifaces.bpf.unstripped.o with a custom command
FAILED: src/core/bpf/restrict_ifaces/restrict-ifaces.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/restrict_ifaces/restrict-ifaces.bpf.c -o src/core/bpf/restrict_ifaces/restrict-ifaces.bpf.unstripped.o
In file included from /usr/include/linux/bpf.h:11,
                 from ../src/core/bpf/restrict_ifaces/restrict-ifaces.bpf.c:6:
/usr/include/linux/types.h:5:10: fatal error: asm/types.h: No such file or directory
    5 | #include <asm/types.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
[4/2361] Generating src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o with a custom command
FAILED: src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/restrict_fs/restrict-fs.bpf.c -o src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o
In file included from ../src/core/bpf/restrict_fs/restrict-fs.bpf.c:9:
/usr/include/linux/types.h:5:10: fatal error: asm/types.h: No such file or directory
    5 | #include <asm/types.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
[19/2361] Generating version.h with a custom command
ninja: build stopped: subcommand failed.
```

**Comment by mbiebl** [comment]
would it make sense to hook up this feature in the CI so it is at least build tested?

**Comment by DaanDeMeyer** [comment]
@mbiebl If we get this working you could update debian's packaging to enable it if the upstream profile is selected and we'll automatically start building it in mkosi.

**Comment by mbiebl** [comment]
@DaanDeMeyer yeah, we could do that

**Comment by jameshilliard** [comment]
> I tried installing gcc-bpf-13 from https://tracker.debian.org/pkg/gcc-bpf

Is that at least version `13.1.0` or later?

**Comment by jameshilliard** [comment]
> Dropping the build flag `-mkernel=5.2` results in

Yeah, so it looks like that flag was [removed](https://github.com/gcc-mirror/gcc/commit/da445a5858299ed2a72af1089c225a438ab93ce2).

**Comment by bluca** [comment]
Removing that is not enough to make it work though, see paste above

**Comment by jameshilliard** [comment]
Hmm, is the `#include <linux/types.h>` header in the bpf programs actually required?

**Comment by mbiebl** [comment]
clang build:
```
[1/564] /usr/bin/clang -std=gnu11 -Wno-compare-distinct-pointer-types -fno-stack-protector -O2 -target bpf -g -c -D__x86_64__ -I. -isystem /usr/include/x86_64-linux-gnu -idirafter /usr/include ../src/core/bpf/restrict_fs/restrict-fs.bpf.c -o src/core/bpf/restrict_fs/restrict-fs.bpf.unstripped.o
```
gcc-bpf build:
```
[2/2390] /usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/socket_bind/socket-bind.bpf.c -o src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o
FAILED: src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o 
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -idirafter /usr/include ../src/core/bpf/socket_bind/socket-bind.bpf.c -o src/core/bpf/socket_bind/socket-bind.bpf.unstripped.o
In file included from ../src/core/bpf/socket_bind/socket-bind-api.bpf.h:9,
                 from ../src/core/bpf/socket_bind/socket-bind.bpf.c:9:
/usr/include/linux/types.h:5:10: fatal error: asm/types.h: No such file or directory
    5 | #include <asm/types.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
```

Notice the missing ` -isystem /usr/include/x86_64-linux-gnu`

**Comment by jameshilliard** [comment]
> Notice the missing ` -isystem /usr/include/x86_64-linux-gnu`

Does it build if you remove the [`and bpf_compiler == 'clang'`](https://github.com/systemd/systemd/blob/v255/meson.build#L1728) check here so that `-isystem` is included?

**Comment by mbiebl** [comment]
Using
```
$ git diff
diff --git a/meson.build b/meson.build
index 8d1cd8a9ed..f484d48601 100644
--- a/meson.build
+++ b/meson.build
@@ -1693,7 +1693,6 @@ if conf.get('BPF_FRAMEWORK') == 1
                 '-std=gnu11',
                 '-fno-stack-protector',
                 '-O2',
-                '-mkernel=5.2',
                 '-mcpu=v3',
                 '-mco-re',
                 '-gbtf',
@@ -1742,7 +1741,7 @@ if conf.get('BPF_FRAMEWORK') == 1
 
         bpf_o_unstripped_cmd += ['-I.']
 
-        if not meson.is_cross_build() and bpf_compiler == 'clang'
+        if not meson.is_cross_build() and bpf_compiler == 'gcc'
                 target_triplet_cmd = run_command('gcc', '-dumpmachine', check: false)
                 if target_triplet_cmd.returncode() == 0
                         target_triplet = target_triplet_cmd.stdout().strip()
```
it builds

**Comment by mbiebl** [comment]
After addressing the build failure with the above changes, I ran the package through autopkgtest.
It failed in TEST-62-RESTRICT-IFACE unfortunately.

an excerpt from the logs/journal:
```
Mär 21 11:01:44 H systemd[1]: libbpf: elf: skipping section(3) .data (size 0)
Mär 21 11:01:44 H systemd[1]: libbpf: elf: skipping unrecognized data section(8) .comment
Mär 21 11:01:44 H systemd[1]: libbpf: prog 'restrict_filesystems': missing .BTF.ext function info for the main program, skipping all of .BTF.ext func info.
Mär 21 11:01:44 H systemd[1]: libbpf: prog 'restrict_filesystems': missing .BTF.ext line info for the main program, skipping all of .BTF.ext line info.
Mär 21 11:01:44 H systemd[1]: libbpf: prog 'restrict_filesystems': BPF program load failed: Permission denied
Mär 21 11:01:44 H systemd[1]: libbpf: prog 'restrict_filesystems': -- BEGIN PROG LOAD LOG --
Mär 21 11:01:44 H systemd[1]: 0: R1=ctx(off=0,imm=0) R10=fp0
Mär 21 11:01:44 H systemd[1]: 0: (62) *(u32 *)(r10 -20) = 0         ; R10=fp0 fp-24=0000????
Mär 21 11:01:44 H systemd[1]: 1: (bf) r0 = r1                       ; R0_w=ctx(off=0,imm=0) R1=ctx(off=0,imm=0)
Mär 21 11:01:44 H systemd[1]: 2: (79) r1 = *(u64 *)(r1 +8)          ; R1_w=scalar()
Mär 21 11:01:44 H systemd[1]: 3: (79) r3 = *(u64 *)(r0 +0)
Mär 21 11:01:44 H systemd[1]: func 'bpf_lsm_file_open' arg0 has btf_id 651 type STRUCT 'file'
Mär 21 11:01:44 H systemd[1]: 4: R0_w=ctx(off=0,imm=0) R3_w=trusted_ptr_file(off=0,imm=0)
Mär 21 11:01:44 H systemd[1]: 4: (67) r1 <<= 32                     ; R1_w=scalar(smax=9223372032559808512,umax=18446744069414584320,smin32=0,smax32=umax32=0,var_off=(0x0; 0xffffffff00000000))
Mär 21 11:01:44 H systemd[1]: 5: (bf) r6 = r1                       ; R1_w=scalar(id=1,smax=9223372032559808512,umax=18446744069414584320,smin32=0,smax32=umax32=0,var_off=(0x0; 0xffffffff00000000)) R6_w=scalar(id=>
Mär 21 11:01:44 H systemd[1]: 6: (c7) r6 s>>= 32                    ; R6_w=scalar(smin=-2147483648,smax=2147483647)
Mär 21 11:01:44 H systemd[1]: 7: (55) if r1 != 0x0 goto pc+49       ; R1_w=0
Mär 21 11:01:44 H systemd[1]: 8: (b7) r2 = 168                      ; R2_w=168
Mär 21 11:01:44 H systemd[1]: 9: (bf) r1 = r10                      ; R1_w=fp0 R10=fp0
Mär 21 11:01:44 H systemd[1]: 10: (0f) r3 += r2                     ; R2_w=168 R3_w=trusted_ptr_file(off=168,imm=0)
Mär 21 11:01:44 H systemd[1]: 11: (07) r1 += -8                     ; R1_w=fp-8
Mär 21 11:01:44 H systemd[1]: 12: (b7) r2 = 8                       ; R2_w=8
Mär 21 11:01:44 H systemd[1]: 13: (85) call bpf_probe_read_kernel#113       ; R0=scalar() fp-8=mmmmmmmm
Mär 21 11:01:44 H systemd[1]: 14: (b7) r4 = 40                      ; R4_w=40
Mär 21 11:01:44 H systemd[1]: 15: (b7) r2 = 8                       ; R2_w=8
Mär 21 11:01:44 H systemd[1]: 16: (79) r3 = *(u64 *)(r10 -8)        ; R3_w=scalar() R10=fp0 fp-8=mmmmmmmm
Mär 21 11:01:44 H systemd[1]: 17: (bf) r1 = r10                     ; R1_w=fp0 R10=fp0
Mär 21 11:01:44 H systemd[1]: 18: (0f) r3 += r4                     ; R3_w=scalar() R4_w=40
Mär 21 11:01:44 H systemd[1]: 19: (07) r1 += -8                     ; R1_w=fp-8
Mär 21 11:01:44 H systemd[1]: 20: (85) call bpf_probe_read_kernel#113       ; R0_w=scalar() fp-8=mmmmmmmm
Mär 21 11:01:44 H systemd[1]: 21: (b7) r5 = 96                      ; R5_w=96
Mär 21 11:01:44 H systemd[1]: 22: (b7) r2 = 8                       ; R2_w=8
Mär 21 11:01:44 H systemd[1]: 23: (79) r3 = *(u64 *)(r10 -8)        ; R3_w=scalar() R10=fp0 fp-8=mmmmmmmm
Mär 21 11:01:44 H systemd[1]: 24: (bf) r1 = r10                     ; R1_w=fp0 R10=fp0
Mär 21 11:01:44 H systemd[1]: 25
Mär 21 11:01:44 H systemd[1]: libbpf: prog 'restrict_filesystems': failed to load: -13
Mär 21 11:01:44 H systemd[1]: libbpf: failed to load object 'restrict_fs_bpf'
Mär 21 11:01:44 H systemd[1]: libbpf: failed to load BPF skeleton 'restrict_fs_bpf': -13
Mär 21 11:01:44 H systemd[1]: bpf-lsm: Failed to load BPF object: Permission denied
```
Full logs attached:
[journal.txt](https://github.com/systemd/systemd/files/14693571/journal.txt)
[log.txt](https://github.com/systemd/systemd/files/14693572/log.txt)

Should I file this as a separate issue?

**Comment by bluca** [comment]
Can you send a PR with the build fixes, as a starting point?

**Comment by bluca** [comment]
@jameshilliard building with gcc-bpf fails again, while it works with clang.

```
/usr/bin/bpf-gcc -std=gnu11 -fno-stack-protector -fno-ssa-phiopt -O2 -mcpu=v3 -mco-re -gbtf -c -D__x86_64__ -mlittle-endian -I. -isystem /usr/include/x86_64-linux-gnu -idirafter /usr/include ../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c -o src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.unstripped.o -I/tmp/cl
In file included from ../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c:20:
../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c: In function 'validate_inode_on_mount':
../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c:80:44: error: invalid first argument format for enum value builtin
   80 |         m = bpf_rdonly_cast(real_mount(v), bpf_core_type_id_kernel(struct mount));
      |                                            ^~~~~~~~~~~~~~~~~~~~~~~
../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c: In function '____userns_restrict_free_user_ns':
../b/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c:168:34: error: invalid first argument format for enum value builtin
  168 |                                  bpf_core_type_id_kernel(struct user_namespace));
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~

```

It's this line:

https://github.com/systemd/systemd/blob/main/src/nsresourced/bpf/userns_restrict/userns-restrict.bpf.c#L80

That can be simplified to this snippet:

`__builtin_btf_type_id(*(typeof(struct mount) *)0, 1);`

Can also be reproduced with godbolt:

https://godbolt.org/z/rfs9TW7Ts

With clang:

https://godbolt.org/z/nWcc8vfGd

gcc-bpf is from git snapshot 2024-03-30

**Comment by cupermir** [comment]
GCC builtins implementation, due to internal reasons is not fully compatible to clang builtin implementation.
For GCC the C pre-processor macro expansion for `bpf_core_type_id_kernel(struct mount))` should not result in `__builtin_btf_type_id(*(typeof(struct mount) *)0, 1)`.
I presume those changes in bpf-next were still not backported in the version of `tools/lib/bpf/bpf_core_read.h` (in kernel) that you rely on. This file contains the CO-RE related macros.

For reference, here is the snippet of the macro definition that makes the differentiation:
```
#ifdef __clang__
#define ___bpf_typeof(type) ((typeof(type) *) 0)
#else
#define ___bpf_typeof1(type, NR) ({					    \
	extern typeof(type) *___concat(bpf_type_tmp_, NR);		    \
	___concat(bpf_type_tmp_, NR);					    \
})
#define ___bpf_typeof(type) ___bpf_typeof1(type, __COUNTER__)
#endif

#define bpf_core_type_id_local(type)					    \
	__builtin_btf_type_id(*___bpf_typeof(type), BPF_TYPE_ID_LOCAL)
```

In bpf-next you can find the commit [12bbcf8](https://kernel.googlesource.com/pub/scm/linux/kernel/git/bpf/bpf-next/+/12bbcf8e840f40b82b02981e96e0a5fbb0703ea9) that adds the support for GCC.

Hope you can try with the proper macros, and confirm if it works right. :smile:

**Comment by bluca** [comment]
Ok that's only available since libbpf 1.4.0, so I'll bump the dependency then https://github.com/libbpf/libbpf/commit/b19fdbf1be21a28f88740375a575ebd9dfbea68f

**Comment by bluca** [comment]
And yes it can compile successfully with a newer libbpf
